docs/advanced-recipes/atom-creators.mdx

```diff
@@ -356,3 +356,66 @@ const ListOfThings = () => {
 ```
 
 <CodeSandbox id="ugkzm0" />
+
+## atomWithDebounce
+
+> `atomWithDebounce` helps with creating an atom where state set should be debounced.
+
+This can be especially useful for text search inputs, where you would like to fire an action only after the user has stopped typing for a delay, instead of firing an action on every keystroke.
+
+```tsx
+export function atomWithDebounce<Value>(
+  initialValue: Value,
+  delayMilliseconds = 500,
+  shouldDebounceOnReset = false
+) {
+  let timeout: NodeJS.Timeout
+
+  // DO NOT EXPORT currentValueAtom as using this atom to set state can cause
+  // inconsistent state between currentValueAtom and debouncedValueAtom
+  const currentValueAtom = atom(initialValue)
+
+  const debouncedValueAtom = atom(
+    initialValue,
+    (get, set, update: SetStateAction<Value>) => {
+      clearTimeout(timeout)
+      const prevValue = get(currentValueAtom)
+      const nextValue =
+        typeof update === 'function'
+          ? (update as (prev: Value) => Value)(prevValue)
+          : update
+
+      set(currentValueAtom, nextValue)
+
+      if (!shouldDebounceOnReset && nextValue === initialValue) {
+        set(debouncedValueAtom, nextValue)
+        return
+      }
+
+      set(currentValueAtom, nextValue)
+
+      timeout = setTimeout(() => {
+        set(debouncedValueAtom, nextValue)
+      }, delayMilliseconds)
+    }
+  )
+
+  const isDebouncingAtom = atom(
+    (get) => get(currentValueAtom) !== get(debouncedValueAtom)
+  )
+
+  return {
+    readOnlyCurrentValueAtom: atom((get) => get(currentValueAtom)),
+    isDebouncingAtom,
+    debouncedValueAtom,
+  }
+}
+```
+
+### Example Usage
+
+The sandbox link below shows how we would use a derived atom to fetch state based on the value of `debouncedValue`.
+
+When typing a pokemon's name in `<SearchInput>`, we do not fire a fetch on every letter, but only after `delayMilliseconds` has passed since the last text input.
```

Comment:
We should explore `useDefferedValue` and `useTransition` for this kind of use cases. It should be far better. This example gives me a little hesitation to merge because it essentially guides to wrong direction.

Do you have another example?

Reply:
May i clarify if `useDefferedValue` and `useTransition` are custom hooks to get atom setters for debouncing the value and setting `isDebouncing`?

Reply:
It's different from debouncing. It will allow data fetching in a lower priority task, so we can keep typing in the text field. The benefit is we don't need to specify the `delay` number. React will automatically adjust it depending on the cpu availability.
This is still a new practice in React, but essentially this will replace the debouncing use case.

Reply:
So, I think we can merge this PR, with a note about new practice in React 18. And, we should work on a new guide about the practice. Hope you can help on it too.

Reply:
I think your use of `useEffect` and `useDefferedValue` is not correct. But, I'm not sure if this use case can be solved with `useDefferedValue`.
I tried with `useTransition`, but still doesn't work as expected. https://codesandbox.io/s/brave-albattani-ldj47c

Hope I can try it again later.

Reply:
Thanks! Good example, I was unfamiliar with that `transition` api as well.

We still don't have control over the number of times the expensive calculation should be done, hence there is some benefit in having a specified timeout delay. The only con would be that this delay sets a minimum threshold before the calculations can happen and is also arbitrarily set.

Reply:
> Hope I can try it again later.

https://codesandbox.io/s/dawn-forest-2wrz9y

Okay, this ☝️ is what I got. We can't stop the big sleep because JS is single threaded. What works is something like this.

Another trial with `useDefferedValue`: https://codesandbox.io/s/gracious-wind-vp54iy?file=/src/_app.js Hmm, not working. I might be misunderstanding something.



> I think there is still some good use-cases for `atomWithDebounce` in react 18.

Yes, overall I agree with it. It's trade-off between automatic and manual. Both are valid.

I'm still interested in solutions with useDefferedValue and useTransition, but this PR is just fine without them.

Reply:
Thanks! Tested your example vs the performance with debounce, and read [this](https://github.com/reactwg/react-18/discussions/41) explanation.

I felt that this author's explanation was the best at exemplifying the cons of `atomWithDebounce`. Using transitions makes the ui responsive, and does not block interaction with the page.

> Another important difference is that a large screen update inside a setTimeout will still lock up the page, just later after the timeout. If the user is still typing or interacting with the page when the timeout fires, they will still be blocked from interacting with the page. But state updates marked with startTransition are interruptible, so they won’t lock up the page. They lets the browser handle events in small gaps between rendering different components. If the user input changes, React won’t have to keep rendering something that the user is no longer interested in.

Will rephrase the doc to explicitly state that the use case for `atomWithDebounce` would be to limit number of function calls derived atoms from the debounced value will make for updates that frequently happen.

Reply:
[5ac4625](https://github.com/pmndrs/jotai/pull/1338/commits/5ac462556b0871b64c44e839203d902f8ffd98ed)